planning/autoware_path_generator/src/node.cpp

```diff
@@ -0,0 +1,418 @@
+      const double lane_length = lanelet::geometry::length(lanelet_sequence.centerline2d());
+      s_end = std::min(s_end, lane_length);
+    }
+
+    if (std::any_of(
+          planner_data_.goal_lanelets.begin(), planner_data_.goal_lanelets.end(),
+          [&](const auto & goal_lanelet) {
+            return lanelet_sequence.lanelets().back().id() == goal_lanelet.id();
+          })) {
+      const auto goal_arc_coordinates =
+        lanelet::utils::getArcCoordinates(lanelet_sequence.lanelets(), planner_data_.goal_pose);
+      s_end = std::min(s_end, goal_arc_coordinates.length);
+    }
+
+    return s_end;
+  }();
+
+  auto path = generate_path(lanelet_sequence, s_start, s_end, params);
+  if (!path) {
+    return std::nullopt;
+  }
+
+  path = autoware::motion_utils::resamplePath(
+    *path, params.input_path_interval, params.enable_akima_spline_first);
+
+  const auto current_seg_idx =
+    autoware::motion_utils::findFirstNearestSegmentIndexWithSoftConstraints(
+      path->points, current_pose, params.ego_nearest_dist_threshold,
+      params.ego_nearest_yaw_threshold);
+
+  path->points = autoware::motion_utils::cropPoints(
+    path->points, current_pose.position, current_seg_idx, params.forward_path_length,
+    params.backward_path_length + params.input_path_interval);
+
+  return path;
+}
+
+std::optional<PathWithLaneId> PathGenerator::generate_path(
+  const lanelet::LaneletSequence & lanelet_sequence, const double s_start, const double s_end,
+  const Params & params) const
+{
+  auto path_points = generate_path_points(lanelet_sequence, s_start, s_end, params);
+  if (path_points.empty()) {
+    return std::nullopt;
+  }
+
+  // append a point if having only one point so that yaw calculation would work
+  if (path_points.size() == 1) {
+    const auto & lane_id = path_points.front().lane_ids.front();
+    const auto & lanelet = planner_data_.lanelet_map_ptr->laneletLayer.get(lane_id);
+    const auto & point = path_points.front().point.pose.position;
+    const auto lane_yaw = lanelet::utils::getLaneletAngle(lanelet, point);
+
+    PathPointWithLaneId path_point{};
+    path_point.lane_ids.push_back(lane_id);
+    constexpr double ds = 0.1;
+    path_point.point.pose.position.x = point.x + ds * std::cos(lane_yaw);
+    path_point.point.pose.position.y = point.y + ds * std::sin(lane_yaw);
+    path_point.point.pose.position.z = point.z;
+    path_points.push_back(path_point);
+  }
+
+  // set yaw to each point
+  for (auto it = path_points.begin(); it != std::prev(path_points.end()); ++it) {
+    const auto angle = autoware::universe_utils::calcAzimuthAngle(
+      it->point.pose.position, std::next(it)->point.pose.position);
+    it->point.pose.orientation = autoware::universe_utils::createQuaternionFromYaw(angle);
+  }
+  path_points.back().point.pose.orientation =
+    std::prev(path_points.end(), 2)->point.pose.orientation;
```

Comment:
will be unnecessary with Autoware Trajectory
https://autowarefoundation.github.io/autoware.universe/main/common/autoware_trajectory/